### PR TITLE
gitea/forgejo: create static gzip and brotli files

### PR DESCRIPTION
**Pull request by Izorkin**
###### Description of changes
Create static gzip and brotli files.
Recommended for use with parameters `services.nginx.recommendedGzipSettings` and `services.nginx.recommendedBrotliSettings`.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/version-management/gitea/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/version-management/gitea/default.nix:60]: This can take a good 5 minutes to run and if the settings in nginx are disabled it is a complete waist of time. Can we move this into passthru and make the gitea nixos module pick it, if one of the recommneded settings is on?

/cc @Ma27

Comment (Izorkin) [reply pkgs/applications/version-management/gitea/default.nix:60]: Add optionn `withStatic`?

**Review by Ma27 — CHANGES_REQUESTED** [review]
I'm fine with this to be an opt-in feature, but I don't want to have it on by default: gitea is supposed to be a somewhat minimal thing to deploy, so I don't want to have static files in the out path for **several** compression algorithms in the runtime closure.

I'm fine with either an option that's **not** on by default or separate derivation outputs (although I guess Sandro won't like that because it implies an increased build-time by default).

> Add option withStatic?

That one is IMHO pretty misleading since I'd expect a `withStatic` to have something to do with static linking.

I'd suggest something like `withGzippedArtifacts`/`withBrotliArtifacts` (though I'm wondering if there's an even better choice than *Artifacts.

Also, speaking of closure-size: would it be an option to only include the compressed artifacts in the output (of gzip xor brotli) and remove the static files? Most likely not by default, though.

**Comment by Izorkin** [comment]
> I'd suggest something like `withGzippedArtifacts`/`withBrotliArtifacts` (though I'm wondering if there's an even better choice than *Artifacts.

Can choose this `withPreCompress` option.

> Also, speaking of closure-size: would it be an option to only include the compressed artifacts in the output (of gzip xor brotli) and remove the static files? Most likely not by default, though.

Probably shouldn't delete the files, because if the client does not support compression, there will be a 404 error.

**Comment by Izorkin** [comment]
Allow this variant? :)
```diff
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index 05ffb5c2b6e..c0defe20723 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -7742,8 +7742,10 @@ with pkgs;
   git-latexdiff = callPackage ../tools/typesetting/git-latexdiff { };

   gitea = callPackage ../applications/version-management/gitea { };
+  giteaNginx = callPackage ../applications/version-management/gitea { withPreCompress = true; };

   forgejo = callPackage ../applications/version-management/forgejo {};
+  forgejoNginx = callPackage ../applications/version-management/forgejo { withPreCompress = true; };

   gokart = callPackage ../development/tools/gokart { };

```

**Review by emilylange — CHANGES_REQUESTED** [review]

File: pkgs/applications/version-management/forgejo/default.nix
Comment (Izorkin) [reply pkgs/applications/version-management/forgejo/default.nix]: Fixed!

File: pkgs/applications/version-management/forgejo/default.nix
Comment (Izorkin) [reply pkgs/applications/version-management/forgejo/default.nix:66]: Not all browsers support brotli compression, in which case gzip compression works.

File: pkgs/applications/version-management/forgejo/default.nix
Comment (IndeedNotJames) [inline pkgs/applications/version-management/forgejo/default.nix:63]: (non-blocking): Could be written as for-loop, but it doesn't really matter.
If you prefer your version, ignore this suggestion :)

```suggestion
     for file in $(find public -type f -regextype posix-extended -iregex '.*\.(css|html|js|svg|ttf|txt)'); do
```

File: pkgs/applications/version-management/forgejo/default.nix
Comment (IndeedNotJames) [inline pkgs/applications/version-management/forgejo/default.nix:65]: Both `gzip` and `brotli` append their corresponding extensions (`.gz` and `.br`) by default.
Any particular reason you used `-n -c >` and `-f -o` instead?

Furthermore, both `gzip` and `brotli` provide the same `-<compression level int>` and `--best` flag.
Why did you go with `-9` (same as `--best`) for `gzip` but `--best` for `brotli`?

It could be simplified to

```suggestion
       gzip --best $file
       brotli --best $file
```

**Review by emilylange — CHANGES_REQUESTED** [review]

File: pkgs/applications/version-management/forgejo/default.nix
Comment (Izorkin) [reply pkgs/applications/version-management/forgejo/default.nix:65]: By default gzip deletes the source file.
I was recommended to add the `-n` option in one of the mastodon/peertube PRs. Now I can't find this place.

I'll change it to this one:
```
  gzip --best --keep $file
  brotli --best --keep $file
```

File: pkgs/applications/version-management/forgejo/default.nix
Comment (IndeedNotJames) [reply pkgs/applications/version-management/forgejo/default.nix]: Thanks for sharing!

I still lean towards dropping that allowlist, though.

Yes, the nixos module for nginx has that set of mime types configured, but users might extend that or use an entirely different web server such as Caddy, which can serve precompressed files as well.

@Ma27 what do you think?

Comment (Izorkin) [reply pkgs/applications/version-management/forgejo/default.nix]: `.woff2` fonts do not compress well.

Comment (Izorkin) [reply pkgs/applications/version-management/forgejo/default.nix]: and images in `.png` format.

File: pkgs/applications/version-management/gitea/default.nix
Comment (SuperSandro2000) [reply pkgs/applications/version-management/gitea/default.nix]: I don't think this should be an option either because we don't need to recompile everything. It is an extra step that is good suited for passthru to extend the already built package but I didn't read the entire thread yet.

**Review by SuperSandro2000 — COMMENTED** [review]
Right now this is rather unattractive. If you turn this on or we even bake it into the nixos module then people need to compile the software **and** compress the assets.

We already have the ``data`` output, why not move the entire postInstall step into passthru.data-compressed? Then we save one unnecessary recompilation of the software. In there we pass in the data output of the package itself, compress it and copy it to $out.

If we are going to start with precompression for more software then we want to do it like this because other software like grafana also runs tests and it takes a good 30m usually for me to recompile and compress.

Also my 2 cents on the option name: ``precompressAssets``.

**Comment by Izorkin** [comment]
> Also my 2 cents on the option name: `precompressAssets`.

Replaced to `preCompressAssets`.

> We already have the `data` output, why not move the entire postInstall step into passthru.data-compressed? Then we save one unnecessary recompilation of the software. In there we pass in the data output of the package itself, compress it and copy it to $out.

Couldn't implement this variant, I did not have enough programming skills

**Comment by SuperSandro2000** [comment]
I took the freedom and force pushed what I had in mind.

Also tested on my gitea and works as expected

File: pkgs/applications/version-management/forgejo/default.nix
Comment (ambroisie) [inline pkgs/applications/version-management/forgejo/default.nix:81]: Why not use `find ... -exec gzip --best --keep {} ';' -exec brotli --best --keep {} ';' ` instead of the loop?  (or even `+` instead of `';'` to reduce the amount of invocations).

File: pkgs/applications/version-management/forgejo/default.nix
Comment (IndeedNotJames) [inline pkgs/applications/version-management/forgejo/default.nix:83]:
```suggestion
        brotli --best --keep --no-copy-stat $file
```

To get rid of all the
```
setting group failed for [/nix/store/jkrq0xdlw50ga14380v3qjlj0lm0rq5x-data-compressed/public/css/dropzone.0142b04d.css.br]: Invalid argument
setting user failed for [/nix/store/jkrq0xdlw50ga14380v3qjlj0lm0rq5x-data-compressed/public/css/dropzone.0142b04d.css.br]: Invalid argument
```
warnings

File: pkgs/applications/version-management/forgejo/default.nix
Comment (IndeedNotJames) [inline pkgs/applications/version-management/forgejo/default.nix:4]: See https://github.com/NixOS/nixpkgs/pull/218171#discussion_r1124686628

**Review by emilylange — CHANGES_REQUESTED** [review]

File: pkgs/applications/version-management/forgejo/default.nix
Comment (Izorkin) [reply pkgs/applications/version-management/forgejo/default.nix:83]: Thanks, fixed.

File: pkgs/applications/version-management/forgejo/default.nix
Comment (Izorkin) [reply pkgs/applications/version-management/forgejo/default.nix:4]: Fixed.

File: pkgs/applications/version-management/forgejo/default.nix
Comment (Izorkin) [reply pkgs/applications/version-management/forgejo/default.nix:81]: Build error with this variant:
```
find -L $out -type f -regextype posix-extended -iregex '.*\.(css|html|js|svg|ttf|txt)' \
  -exec gzip --best --keep --force {} ';' \
  -exec brotli --best --keep --no-copy-stat {}
```
```
error: builder for '/nix/store/sjain6jfba0ipjwamx15a63x3460wlz1-data-compressed.drv' failed with exit code 1;
       last 10 log lines:
       > /nix/store/b2gjap9zxk3cfcys43caa4hns85gf1nq-forgejo-1.18.5-0-data/templates/repo/find:
       > /nix/store/b2gjap9zxk3cfcys43caa4hns85gf1nq-forgejo-1.18.5-0-data/templates/repo/issue:
       > /nix/store/b2gjap9zxk3cfcys43caa4hns85gf1nq-forgejo-1.18.5-0-data/templates/repo/issue/view_content:
       > /nix/store/b2gjap9zxk3cfcys43caa4hns85gf1nq-forgejo-1.18.5-0-data/templates/repo/issue/fields:
       > /nix/store/b2gjap9zxk3cfcys43caa4hns85gf1nq-forgejo-1.18.5-0-data/templates/repo/issue/labels:
       > /nix/store/b2gjap9zxk3cfcys43caa4hns85gf1nq-forgejo-1.18.5-0-data/templates/repo/graph:
       > /nix/store/b2gjap9zxk3cfcys43caa4hns85gf1nq-forgejo-1.18.5-0-data/templates/repo/migrate:
       > /nix/store/b2gjap9zxk3cfcys43caa4hns85gf1nq-forgejo-1.18.5-0-data/templates/explore:
       > /nix/store/b2gjap9zxk3cfcys43caa4hns85gf1nq-forgejo-1.18.5-0-data/templates/base:
       > find: missing argument to `-exec'
       For full logs, run 'nix log /nix/store/sjain6jfba0ipjwamx15a63x3460wlz1-data-compressed.drv'.

```

Comment (Izorkin) [reply pkgs/applications/version-management/forgejo/default.nix]: Fixed.

**Comment by SuperSandro2000** [comment]
~~I tested this on my gitea installation and noticed that patches to the assets through overlays where ignored. Since buildGoModules has no finalAttrs support I went with an older ``let self = ... in self`` cosntruct.~~ 🤦🏼 my patching is just to brittle. Sorry for the noise.

LGTM from me now. Thanks for the patience!

**Review by SuperSandro2000 — APPROVED** [review]

**Review by Ma27 — APPROVED** [review]

**Comment by Ma27** [comment]
OK, just pushed a small change, I think that `data-compressed` is way too undescriptive. DIdn't want to go through yet another review cycle so I just went ahead and pushed. After that I'd merge :)

**Comment by Izorkin** [comment]
In the `forgejo` package `data-compressed` must also be renamed.
Rebase PR?

**Comment by Ma27** [comment]
@ofborg test

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by dotlambda** [comment]
Can someone point me to how this is supposed to be used?

**Comment by Izorkin** [comment]
> Can someone point me to how this is supposed to be used?

Will be used in nginx configuration - https://github.com/NixOS/nixpkgs/pull/218020

**Pull request merged**
